usbsdmux/usbsdmux.py

```diff
@@ -18,12 +18,56 @@
 # License along with this library; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 
+import os
 import time
 
-from .pca9536 import Pca9536
+from .i2c_gpio import Pca9536, Tca6408
 
 
-class UsbSdMux(object):
+class UsbSdMuxFactory:
+    SD_MUX_TYPE_LEGACY = "legacy"
+    SD_MUX_TYPE_FAST = "fast"
+
+    class UnknownUsbSdMuxRevisionException(Exception):
+        pass
+
+    @staticmethod
+    def get(sg, sd_mux_type=None):
+        """
+        Create a new UsbSdMux.
+
+        Arguments:
+        sg -- /dev/sg* to use
+        sd_mux_type -- None or UsbSdMux.SD_MUX_TYPE_*
+        """
+
+        if not sd_mux_type:
+            # Let's try to detect the type of USB-SD-Mux connected
+            # Get the name of the sg-device from the path
+            base_sg = os.path.realpath(sg)
+            sg_name = os.path.basename(base_sg)
+            try:
+                with open(f"/sys/class/scsi_generic/{sg_name}/device/model", "r") as fh:
+                    model = fh.read().strip()
+                if model == "sdmux HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_LEGACY
+                elif model == "sdFST HS-SD/MMC":
+                    sd_mux_type = UsbSdMuxFactory.SD_MUX_TYPE_FAST
```

Comment:
Instead of indirecting via sd_mux_type, you could just return the matched class here. That would also avoid the need to pass a sd_mux_type` to override. You'd only call this function (perhaps called `autoselect`) if you don't know the type already.

That would simplify the control flow a lot.

Reply:
Sorry. I don't understand what you mean here. Is there something lost in the malformed highlight?

Reply:
I got rid of the type override and now simply return the correct instance.